README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# powerPsych <img src="man/figures/logo.png" align="right" height=230/>

<!-- badges: start -->

![](https://img.shields.io/badge/AndrewLi-STAT545-green) [![Lifecycle:
questioning](https://img.shields.io/badge/lifecycle-questioning-blue.svg)](https://www.tidyverse.org/lifecycle/#questioning)
<!-- badges: end -->

The goal of `powerPsych` is to provide behavioral scientists a package
to calculate and understand effect size and power. `powerPsych` is a
collection of statistical analysis tools, visualization, and calculator
that are useful for research. This package is separated by three main
analysis:

1.  Effect size
2.  Statistical power
3.  Sample size calculator

This package was built for my STAT 545B course at the University of
British Columbia. I built this package based
[this](https://www.ai-therapy.com) website’s calculations. Admittedly,
their website is much easier to use but here I offer an alternative for
people that prefer typesetting.

## Installation

`powerPsych` is not on CRAN, you need to install it via [GitHub]():

``` r
install.github("andr3wli/powerPsych")
```

## Effect size

Effect size is a standardized way to report the strength of an apparent
relationship. This part of the package dives into *Cohen’s d*. Cohen’s d
is very popular in psychology. `cohen.d` is a function that exists
already that computes the Cohen’s d and Hedge’s effect size statistic. I
offer two new functions to help calculate and interpret this statistic.
Cohen’s d can be categorized as follows:

| cohen’s d | Interpretation  |
| --------- | --------------- |
| 0.2       | A small effect  |
| 0.5       | A medium effect |
| 0.8+      | A large effect  |

  - `cohens_d` is a calculator that compares two populations/groups
    effect size via the two means. `cohens_d` consumes the mean and
    standard deviation of of two groups and it calculates the statistic
    and informs the user of the effect size.

  - `cohens_plot` is a visualization tool to help aid the interpretation
    of Cohen’s d value.

### Here are some examples of `cohens_d`:

``` r
library(powerPsych)

cohens_d(0, 0.35, 0.5, 0.6, sd = "diff")
#> Cohen's d is 0.634 which is considered to be medium effect size.
cohens_d(0, -0.35, 0.5, sd = "same")
#> Cohen's d is 0.7 which is considered to be a relatively small effect size.
```

### Here are some examples of `cohens_plot`:

``` r
cohens_plot(0, 1, 0.5, 0.5)
```

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" />

``` r
cohens_plot(0, 1, 0.56, 0.5, overlap = F)
```

<img src="man/figures/README-unnamed-chunk-4-2.png" width="100%" />

**Note:** Effect sizes are computed using the methods outlined in the
paper ""Olejnik, S. & Algina, J. 2003. Generalized Eta and Omega Squared
Statistics: Measures of Effect Size for Some Common Research Designs
*Psychological Methods. 8:(4)434-447"*.

## Workflow

Here I will describe the workflow for building this package.

1.  I loaded the `devtools` package and created a new package called
    powerPsych

<!-- end list -->

  - `devtools::create_package("powerPsych)`

<!-- end list -->

2.  Initialized the package repo

<!-- end list -->

  - `use_git()`

<!-- end list -->

3.  Created my R script for `cohens_d`

<!-- end list -->

  - `use_r("cohens_d")`

<!-- end list -->

4.  Used Roxygen2 to create the documents and to export them

5.  Created my tests for the function

<!-- end list -->

  - `use_testthat()`
  - `use_test(cohens_d)`

<!-- end list -->

6.  Created a README and rendered the R markdown file to a regualr
    markdown file

<!-- end list -->

  - `use_readme_rmd`
  - `build_readme()`

<!-- end list -->

7.  created my vignettes and built it

<!-- end list -->

  - `use_vignette("cohens_d")`
  - `build_vignette()`

## Statistical power

I will build upon this package for the final assignment for STAT 545B. I
plan on having a t-test calculator and a F-test calculator.

## Sample size

I will build upon this package for the final assignment for STAT 545B. I
plan on creating a function that can inform researchers the sample size
needed so that the experiment has sufficient statistical power.

## Acknowledgement

The STAT 545B teaching team deserves a huge amount of credit for
teaching this course so wonderfully\! Thank you to our prof [Vincenzo
Coia](https://vincenzocoia.com) and our wonderful TAs [Almas Khan](),
[Diana Lin](http://www.birollab.ca/member/dlin), [Icíar Fernández
Boyano](), and [Victor Yuan]().

## Code of Conduct

Please note that the powerPsych project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
## powerPsych 0.1.1 (2020-11-19)

In this version of `powerPsych` I am done with effect sizes. Moving forward, I will be making the sample size calculator and power calculator. Here is what's new:

* `cohens_plot()` is here! 
* fixed some spelling mistakes
* added examples to the readme.rmd

## powerPsych 0.1.0 (2020-11-18)

The first version of `powerPsych` is out! Woohoo! It doesn't have many functions (only one) as I had to submit this for an assignment. Take a look at the README.md file to see what I want to do with this package for our final assignment.

* Initial release
* `cohens_d()` is a calculator for Cohen's d


tests/testthat.R
library(testthat)
library(powerPsych)

test_check("powerPsych")


tests/testthat/test-cohens_d.R
test_that("prints the correct calculation and interpretation", {
  expect_that(cohens_d(0, 1.4, 0.6, sd = "same"),
              prints_text("Cohen's d is 2.333 which is considered to be a relatively large effect size."))
  expect_that(cohens_d(0, 1, 0.5, 0.6, sd = "diff"),
              prints_text("Cohen's d is 1.811 which is considered to be a relatively large effect size."))
  expect_that(cohens_d(0, 0.1, 0.5, 0.6, sd = "diff"),
              prints_text("Cohen's d is 0.181 which is considered to be a relatively small effect size."))
  expect_that(cohens_d(0, 0.35, 0.5, 0.6, sd = "diff"),
              prints_text("Cohen's d is 0.634 which is considered to be medium effect size."))
})

test_that("expect error when sd.two is missing when sd = 'diff", {
  expect_error(cohens_d(0, 0.35, 0.5, sd = "diff"),
               "If the standard deviation of the groups are different, you must enter a value for sd.two")
  expect_error(cohens_d(1, 0.35, 0.5, sd = "diff"),
               "If the standard deviation of the groups are different, you must enter a value for sd.two")
})


tests/testthat/test-cohens_plot.R
test_that("Plot returns ggplot object",{
  p <- cohens_plot(0, 1, 0.5, 0.8, overlap = T)
  p2 <- cohens_plot(0, 1, 0.5, 0.8, overlap = F)
  expect_is(p,"ggplot")
  expect_is(p2,"ggplot")
})

# test_that("Plot layers match expectations",{
#   p <- cohens_plot(0, 1, 0.5, 0.8, overlap = T)
#   expect_identical(p$layers[[1]]$geom$objname, "line")
#   expect_identical(p$layers[[2]]$geom$objname, "line")
#   expect_identical(p$layers[[3]]$geom$objname, "vline")
#   expect_identical(p$layers[[4]]$geom$objname, "vline")
#   expect_identical(p$layers[[5]]$geom$objname, "polygon")
# })

# test_that("Plot uses correct data", {
#   p <- cohens_plot(0, 1, 0.5, 0.8, overlap = T)
#   expect_that(df1, equals(p$df1))
#
# })

test_that("y and x axis and title labelled correctly",{
  p <- cohens_plot(0, 1, 0.5, 0.8, overlap = T)
  p2 <- cohens_plot(0.3, 0.5, 0.5, 0.8, overlap = F)
  expect_identical(p$labels$y, "")
  expect_identical(p$labels$x, "")
  expect_identical(p$labels$title, "Cohen's d: 1.499")
  expect_identical(p2$labels$y, "")
  expect_identical(p2$labels$x, "")
  expect_identical(p2$labels$title, "Cohen's d: 0.3")
})


R/cohens_plot.R
#' Calculate and plots Cohen's d given given some descriptive statistic
#'
#' @param mean.one consumes the mean of first group
#' @param mean.two consumes the mean of second group
#' @param sd.one consumes the standard deviation of first group
#' @param sd.two consumes the standard deviation of second group.
#' @param overlap default is set to true. Type false if you do not want an overlap in the plots.
#'
#' @return
#' a ggplot showing the effect for the given mean and standard deviation
#'
#' @export
#'
#' @examples
#' cohens_plot(0, 1, 0.5, 0.8, overlap = FALSE)
#' cohens_plot(0, 1, 0.5, 0.8)
#' cohens_plot(0, 1, 0.5, 0.8, overlap = TRUE)
cohens_plot <- function(mean.one, mean.two, sd.one, sd.two, overlap = TRUE) {

  cd <- (mean.two - mean.one) / sqrt(((sd.one^2) + (sd.two^2)) / 2)

  ES <- 0.8
  mean1 <- ES*1 + 1
  x <- seq(1 - 3*1, mean1 + 3*1, .01)

  # generate normal dist #1
  y1 <- stats::dnorm(x, mean.one, sd.one)
  df1 <- data.frame("x" = x, "y" = y1)

  # generate normal dist #2
  y2 <- stats::dnorm(x, mean.two, sd.two)
  df2 <- data.frame("x" = x, "y" = y2)

  # data for overlap
  y.poly <- pmin(y1,y2)
  poly <- data.frame("x" = x, "y" = y.poly)

  if(overlap == TRUE){
    plot <- ggplot2::ggplot(df1, ggplot2::aes_(x = ~ x, y = ~ y, color="treatment")) +
      ggplot2::geom_line(size=1) +
      ggplot2::geom_line(data=df2, ggplot2::aes(color="control"),size=1) +
      ggplot2::guides(color = FALSE) +
      ggplot2::scale_color_manual(values = c("#b2182b", "#4d4d4d")) +
      ggplot2::labs(x = "", y = "", title = paste("Cohen's d:", round(abs(cd), digits = 3))) +
      ggplot2::geom_vline(xintercept = mean.one, linetype="dotted") +
      ggplot2::geom_vline(xintercept = mean.two, linetype="dotted") +
      ggplot2::geom_polygon(ggplot2::aes(color=NULL), data=poly, fill="#ef8a62", alpha=I(4/10)) +
      # xlim(-7, 7)
      ggplot2::theme_void()+
      ggplot2::theme(aspect.ratio = 4/10)
    return(plot)
  }

  if(overlap == FALSE){
    plot <- ggplot2::ggplot(df1, ggplot2::aes_(x = ~ x, y = ~ y, color="treatment")) +
      ggplot2::geom_line(size=1) +
      ggplot2::geom_line(data=df2, ggplot2::aes(color="control"),size=1) +
      ggplot2::guides(color = FALSE) +
      ggplot2::scale_color_manual(values = c("#b2182b", "#4d4d4d")) +
      ggplot2::labs(x = "", y = "", title = paste("Cohen's d:", round(abs(cd), digits = 3))) +
      ggplot2::geom_vline(xintercept = mean.one, linetype="dotted") +
      ggplot2::geom_vline(xintercept = mean.two, linetype="dotted") +
      # xlim(-7, 7)
      ggplot2::theme_void() +
      ggplot2::theme(aspect.ratio = 4/10)
    return(plot)
  }

}


R/cohens_d.R
#' Calculate Cohen's d given some descriptive statistics
#'
#' @param mean.one consumes the mean of first group
#' @param mean.two consumes the mean of second group
#' @param sd.one consumes the standard deviation of first group
#' @param sd.two consumes the standard deviation of second group. This can be left empty if the groups share the same standard deviation.
#' @param sd Type "same" if both groups have the same sd. Type "diff" if the groups do not have the same standard deviation. In this case, the pooled standard deviation is used.
#'
#' @return
#' A string containing Cohen's d statistic that informs user of effect size
#'
#' @export
#'
#'
#'
#' @examples
#' cohens_d(0, 1.4, 0.6, sd = "same")
#' cohens_d(mean.one = 0, mean.two = 1, sd.one = 0.5, sd.two = 0.6, sd = "diff")
cohens_d <- function(mean.one, mean.two, sd.one, sd.two, sd = c("same", "diff")) {

  if(sd == "same") {
    cd_same <- (mean.two - mean.one) / sd.one

    if(cd_same <= 0.2) {
      return(cat("Cohen's d is", round(abs(cd_same), digits = 3), "which is considered to be a relatively small effect size."))
    }
    if(cd_same > 0.5 & cd_same < 0.8) {
      return(cat("Cohen's d is", round(abs(cd_same), digits = 3), "which is considered to be medium effect size."))
    }
    if(cd_same > 0.8) {
      return(cat("Cohen's d is", round(abs(cd_same), digits = 3), "which is considered to be a relatively large effect size."))
    }
  }


  if(sd == "diff" & missing(sd.two)) {
    stop("If the standard deviation of the groups are different, you must enter a value for sd.two")
}

  if(sd == "diff") {
    cd_diff <- (mean.two - mean.one) / sqrt(((sd.one^2) + (sd.two^2)) / 2)


    if(cd_diff <= 0.2) {
      return(cat("Cohen's d is", round(abs(cd_diff), digits = 3), "which is considered to be a relatively small effect size."))
    }
    if(cd_diff > 0.5 & cd_diff < 0.8) {
      return(cat("Cohen's d is", round(abs(cd_diff), digits = 3), "which is considered to be medium effect size."))
    }
    if(cd_diff > 0.8) {
      return(cat("Cohen's d is", round(abs(cd_diff), digits = 3), "which is considered to be a relatively large effect size."))

    }
  }

}
